test/srt/test_p2p_transfer.py

```diff
@@ -0,0 +1,1010 @@
+    """Create mock weights for a single Qwen3-32B transformer layer."""
+    weights = {}
+    qkv_size = config.hidden_size + 2 * (config.hidden_size // config.num_attention_heads) * config.num_key_value_heads
+    weights["qkv_proj"] = torch.randn(qkv_size, config.hidden_size, device=device, dtype=dtype)
+    weights["o_proj"] = torch.randn(config.hidden_size, config.hidden_size, device=device, dtype=dtype)
+    weights["gate_up_proj"] = torch.randn(config.intermediate_size * 2, config.hidden_size, device=device, dtype=dtype)
+    weights["down_proj"] = torch.randn(config.hidden_size, config.intermediate_size, device=device, dtype=dtype)
+    weights["ln1"] = torch.randn(config.hidden_size, device=device, dtype=dtype)
+    weights["ln2"] = torch.randn(config.hidden_size, device=device, dtype=dtype)
+    return weights
+
+
+def create_qwen3_full_model(config: Qwen3_32BConfig, device: torch.device, dtype=torch.float16) -> Dict[str, torch.Tensor]:
+    """Create full Qwen3-32B model weights (using 8 layers for testing)."""
+    weights = {}
+    weights["embedding"] = torch.randn(config.vocab_size, config.hidden_size, device=device, dtype=dtype)
+
+    num_test_layers = min(config.num_layers, 8)
+    for layer_idx in range(num_test_layers):
+        layer_weights = create_qwen3_layer_weights(config, device, dtype)
+        for name, weight in layer_weights.items():
+            weights[f"layer_{layer_idx}_{name}"] = weight
+
+    weights["lm_head"] = torch.randn(config.vocab_size, config.hidden_size, device=device, dtype=dtype)
+    return weights
+
+
+class TrainingWeightSender:
+    """Training-side weight sender that handles RDMA write operations."""
+
+    def __init__(self, hostname: str, port: int, gpu_id: int):
+        self.hostname = hostname
+        self.port = port
+        self.gpu_id = gpu_id
+        self.running = False
+
+        self.engine = MooncakeTransferEngine(
+            hostname=hostname,
+            gpu_id=gpu_id,
+            ib_device=None,
+        )
+
+        self.context = zmq.Context()
+        self.router_socket = self.context.socket(zmq.ROUTER)
+        self.router_socket.bind(f"tcp://{hostname}:{port}")
+
+        self.weight_buffers = {}
+
+        logger.info(f"[TrainingSender] Initialized on {hostname}:{port}, GPU {gpu_id}")
+
+    def register_weights(self, weights_dict: Dict[str, torch.Tensor]):
+        """Register all weight tensors for sending."""
+        for name, tensor in weights_dict.items():
+            ptr = tensor.data_ptr()
+            length = tensor.numel() * tensor.element_size()
+
+            if name in self.weight_buffers:
+                existing_ptr = self.weight_buffers[name]['ptr']
+                if existing_ptr == ptr:
+                    self.weight_buffers[name]['tensor'] = tensor
+                    continue
+
+            self.engine.register(ptr, length)
```

Comment:
I see you call this method twice. Would engine reuse its cache and skip the registration?